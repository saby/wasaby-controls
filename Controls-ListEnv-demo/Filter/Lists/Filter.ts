function filter(item, filter: { department: string[] | string; city: string[] | string }): boolean {
    if (filter.department) {
        return Array.isArray(filter.department)
            ? filter.department.includes(item.get('department'))
            : filter.department.toLowerCase() === item.get('department').toLowerCase();
    }
    if (filter.city?.length) {
        return Array.isArray(filter.city)
            ? filter.city.includes(item.get('city'))
            : filter.city.toLowerCase() === item.get('city').toLowerCase();
    }
    return true;
}

filter._moduleName = 'Controls-ListEnv-demo/Filter/Lists/Filter';

export = filter;
