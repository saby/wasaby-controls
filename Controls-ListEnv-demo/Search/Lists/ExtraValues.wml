<div class="controlsDemo__wrapper">
    <Controls-ListEnv.searchConnected:Input storeId="search"/>
    <div class="ws-flexbox">
        <Controls.list:View storeId="list1"/>
        <Controls.list:View storeId="list2"/>
    </div>
</div>
