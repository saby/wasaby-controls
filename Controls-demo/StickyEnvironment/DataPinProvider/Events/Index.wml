<div class="controls-demo_Wrapper controlsDemo_fixedWidth500">
   <div data-qa="controls-demo__text_top">Top edge data id: {{_topEdgeData}}</div>
   <div data-qa="controls-demo__text_bottom">Bottom edge data id: {{_bottomEdgeData}}</div>

   <Controls.stickyEnvironment:DataPinProvider on:edgesDataChanged="_onEdgesDataChanged()">
      <Controls.scroll:Container attr:style="height: 500px;">
         <Controls.list:View
            source="{{_viewSource}}"
            keyProperty="key"
            navigation="{{_navigation}}"
         >
            <ws:virtualScrollConfig page="{{50}}" pageSize="{{50}}"/>
            <ws:itemTemplate>
               <ws:partial template="Controls/list:ItemTemplate">
                  <ws:contentTemplate>
                     <Controls.stickyEnvironment:DataPinContainer data="{{itemTemplate.item.contents.get('key')}}">
                        <strong>Запись с id="{{itemTemplate.item.contents.get('key')}}"</strong>
                     </Controls.stickyEnvironment:DataPinContainer>

                     {{itemTemplate.item.contents.get('title')}}
                     <i data-qa="Controls-Demo_StickyEnvironment__hooks">h</i>
                  </ws:contentTemplate>
               </ws:partial>
            </ws:itemTemplate>
         </Controls.list:View>
      </Controls.scroll:Container>
   </Controls.stickyEnvironment:DataPinProvider>
</div>
