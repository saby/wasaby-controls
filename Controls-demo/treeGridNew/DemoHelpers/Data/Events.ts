import { IColumn } from 'Controls/grid';
import { IData } from '../Interface';

export const Events = {
    getColumns: (): IColumn[] => {
        return [
            {
                displayProperty: 'title',
                width: '',
            },
        ];
    },
    getDataBeforeExpanded: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Асинхронная обработка перед разворотом узла',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Асинхронная обработка перед разворотом узла',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataBeforeCollapsed: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Асинхронная обработка перед сворачиванием узла',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Асинхронная обработка перед сворачиванием узла',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataCollapsedItemsChanged: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Этот узел нельзя свернуть',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Этот узел можно свернуть',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataExpandedItemsChanged: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Этот узел нельзя развернуть',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Этот узел можно развернуть',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
};
