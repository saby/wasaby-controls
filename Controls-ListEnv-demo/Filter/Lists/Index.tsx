import * as React from 'react';
import { dropdownConfig } from 'Controls-ListEnv-demo/FilterPanel/View/Editors/DropdownEditor/Index';
import { lookupBasicConfig } from 'Controls-ListEnv-demo/Filter/View/Editors/LookupEditor/MultiSelect/Index';
import { Memory } from 'Types/source';
import { sourceData } from 'Controls-ListEnv-demo/Filter/resources/DataStorage';
import { View as FilterView } from 'Controls-ListEnv/filterConnected';
import { IColumn, IHeaderCell, View as GridControl } from 'Controls/grid';
import { Title } from 'Controls/heading';
import * as filter from './Filter';
import 'css!Controls-ListEnv-demo/Filter/filter';

const columns: IColumn[] = [
    {
        displayProperty: 'department',
    },
    {
        displayProperty: 'owner',
    },
    {
        displayProperty: 'salary',
    },
    {
        displayProperty: 'city',
    },
    {
        displayProperty: 'date',
    },
];

const header: IHeaderCell[] = [
    { caption: 'Отдел' },
    { caption: 'Руководитель' },
    { caption: 'Зарплата' },
    { caption: 'Город' },
    { caption: 'Дата' },
];

function FilterWithLists(props: unknown, ref: React.ForwardedRef<unknown>): JSX.Element {
    return (
        <div className={'engine-demo__Widgets_scroll'} ref={ref}>
            <div className={'engine-demo__Widgets_filter-container'}>
                <div>
                    <FilterView storeId={'filter'} />
                    <div className={'ws-flexbox'}>
                        <div>
                            <Title caption={'Список, который фильтрует только по подразделению'} />
                            <GridControl storeId={'department'} columns={columns} header={header} />
                        </div>
                        <div>
                            <Title caption={'Список, который фильтрует только по городу'} />
                            <GridControl storeId={'city'} columns={columns} header={header} />
                        </div>
                    </div>
                </div>
            </div>
        </div>
    );
}

const forwardedDemo = React.forwardRef(FilterWithLists);

forwardedDemo.getLoadConfig = function () {
    return {
        filter: {
            dataFactoryName: 'Controls-ListEnv/filterDataFactory:Factory',
            dataFactoryArguments: {
                filterDescription: [...[dropdownConfig, lookupBasicConfig]],
            },
        },
        city: {
            dataFactoryName: 'Controls/dataFactory:List',
            dataFactoryArguments: {
                clearResult: true,
                sliceExtraValues: [
                    {
                        propName: 'filterDescription',
                        dependencyName: 'filter',
                        dependencyPropName: 'filterDescription',
                        prepare: function (filterDescription) {
                            return filterDescription.filter((filterItem) => {
                                return filterItem.name === 'city';
                            });
                        },
                    },
                ],
                loaderExtraValues: [
                    {
                        propName: 'filterDescription',
                        dependencyName: 'filter',
                        dependencyPropName: 'filterDescription',
                        prepare: function (filterDescription) {
                            return filterDescription.filter((filterItem) => {
                                return filterItem.name === 'city';
                            });
                        },
                    },
                ],
                keyProperty: 'department',
                displayProperty: 'title',
                source: new Memory({
                    data: sourceData,
                    keyProperty: 'department',
                    filter,
                }),
            },
        },
        department: {
            dataFactoryName: 'Controls/dataFactory:List',
            dataFactoryArguments: {
                clearResult: true,
                sliceExtraValues: [
                    {
                        propName: 'filterDescription',
                        dependencyName: 'filter',
                        dependencyPropName: 'filterDescription',
                        prepare: function (filterDescription) {
                            return filterDescription.filter((filterItem) => {
                                return filterItem.name === 'department';
                            });
                        },
                    },
                ],
                loaderExtraValues: [
                    {
                        propName: 'filterDescription',
                        dependencyName: 'filter',
                        dependencyPropName: 'filterDescription',
                        prepare: function (filterDescription) {
                            return filterDescription.filter((filterItem) => {
                                return filterItem.name === 'department';
                            });
                        },
                    },
                ],
                keyProperty: 'department',
                displayProperty: 'title',
                source: new Memory({
                    data: sourceData,
                    keyProperty: 'department',
                    filter,
                }),
            },
        },
    };
};

export default forwardedDemo;
