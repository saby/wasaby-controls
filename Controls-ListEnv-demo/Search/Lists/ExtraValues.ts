import { Control, IControlOptions, TemplateFunction } from 'UI/Base';
import { Memory } from 'Types/source';
import * as filter from '../Util/DataFilter';
import * as controlTemplate from 'wml!Controls-ListEnv-demo/Search/Lists/ExtraValues';
import 'css!Controls-ListEnv-demo/Search/Index';

export default class Index extends Control<IControlOptions> {
    protected _template: TemplateFunction = controlTemplate;

    // Т.к. механизм построения демо примеров отличается от механизма построения страницы, то данный способ предзагрузки
    // используется только для демо примеров. Посмотреть как настраивать предзагрузку на странице можно по ссылке
    // https://wi.sbis.ru/doc/platform/developmentapl/interface-development/application-configuration/create-page/accordion/content/prefetch-config/
    static getLoadConfig() {
        return {
            search: {
                dataFactoryName: 'Controls-ListEnv/searchDataFactory:Factory',
                dataFactoryArguments: {},
            },
            list1: {
                dataFactoryName: 'Controls/dataFactory:List',
                dataFactoryArguments: {
                    loaderExtraValues: {
                        searchValue: ['search', 'searchValue'],
                    },
                    sliceExtraValues: {
                        searchValue: ['search', 'searchValue'],
                    },
                    source: new Memory({
                        keyProperty: 'id',
                        data: [
                            {
                                id: 1,
                                title: 'Разработка',
                            },
                            {
                                id: 2,
                                title: 'Продвижение СБИС',
                            },
                            {
                                id: 3,
                                title: 'Федеральная клиентская служба',
                            },
                        ],
                        filter,
                    }),
                    searchParam: 'title',
                    keyProperty: 'id',
                    displayProperty: 'title',
                },
            },
            list2: {
                dataFactoryName: 'Controls/dataFactory:List',
                dataFactoryArguments: {
                    loaderExtraValues: {
                        searchValue: ['search', 'searchValue'],
                    },
                    sliceExtraValues: {
                        searchValue: ['search', 'searchValue'],
                    },
                    source: new Memory({
                        keyProperty: 'id',
                        data: [
                            {
                                id: 4,
                                title: 'Служба эксплуатации',
                            },
                            {
                                id: 5,
                                title: 'Технологии и маркетинг',
                            },
                            {
                                id: 6,
                                title: 'Федеральный центр продаж',
                            },
                        ],
                        filter,
                    }),
                    searchParam: 'title',
                    keyProperty: 'id',
                    displayProperty: 'title',
                },
            },
        };
    }
}
