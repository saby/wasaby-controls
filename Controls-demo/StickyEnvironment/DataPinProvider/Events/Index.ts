import { Memory } from 'Types/source';
import { SyntheticEvent } from 'UI/Vdom';
import { IEdgesData } from 'Controls/stickyEnvironment';
import { Control, IControlOptions, TemplateFunction } from 'UI/Base';
import { generateData } from 'Controls-demo/list_new/DemoHelpers/DataCatalog';
import * as template from 'wml!Controls-demo/StickyEnvironment/DataPinProvider/Events/Index';

export default class Index extends Control<IControlOptions> {
    protected _template: TemplateFunction = template;

    protected _topEdgeData: number;
    protected _bottomEdgeData: number;

    protected _viewSource: Memory;
    protected _navigation: object = {
        source: 'page',
        view: 'infinity',
        sourceConfig: {
            page: 10,
            pageSize: 50,
            hasMore: false,
        },
    };

    private _dataArray: { key: number; title: string }[] = generateData<{
        key: number;
        title: string;
    }>({
        count: 1000,
        entityTemplate: { title: 'lorem' },
    });

    protected _beforeMount(): void {
        this._viewSource = new Memory({
            keyProperty: 'key',
            data: this._dataArray,
        });
    }

    protected _onEdgesDataChanged(
        event: SyntheticEvent,
        data: IEdgesData<number>
    ): void {
        this._topEdgeData = data.top.above || data.top.below;
        this._bottomEdgeData = data.bottom.above;
    }
}
