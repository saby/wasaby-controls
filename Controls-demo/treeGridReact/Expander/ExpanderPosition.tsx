import * as React from 'react';

import { getArgs } from 'UICore/Events';

import 'Controls/gridReact';
import { ItemsView as TreeGridItemsView } from 'Controls/treeGrid';
import { Control as Tumbler } from 'Controls/Tumbler';
import { TExpanderPosition } from 'Controls/baseTree';
import { RecordSet } from 'Types/collection';

import { getBaseColumns, getBaseRecordSet } from '../Data';

function getExpanderPositions(): RecordSet {
    return new RecordSet({
        keyProperty: 'key',
        rawData: [{ key: 'default' }, { key: 'right' }, { key: 'custom' }],
    });
}

interface IProps {
    expanderPosition?: TExpanderPosition;
}

export default React.forwardRef(
    (props: IProps, ref: React.ForwardedRef<HTMLDivElement>) => {
        const columns = React.useMemo(() => {
            return getBaseColumns();
        }, []);
        const items = React.useMemo(() => {
            return getBaseRecordSet();
        }, []);
        const expanderPositions = React.useMemo(() => {
            return getExpanderPositions();
        }, []);

        const [expanderPosition, setExpanderPosition] =
            React.useState<TExpanderPosition>(
                props.expanderPosition || 'default'
            );

        return (
            <div
                ref={ref}
                className={'controlsDemo__wrapper controlsDemo__maxWidth500'}
            >
                <Tumbler
                    keyProperty="key"
                    displayProperty="key"
                    items={expanderPositions}
                    selectedKey={expanderPosition}
                    onSelectedKeyChanged={(event) => {
                        return setExpanderPosition(getArgs(event)[1]);
                    }}
                    customEvents={['onSelectedKeyChanged']}
                />
                <TreeGridItemsView
                    items={items}
                    columns={columns}
                    root={null}
                    keyProperty={'key'}
                    nodeProperty={'type'}
                    parentProperty={'parent'}
                    expanderPosition={expanderPosition}
                />
            </div>
        );
    }
);
