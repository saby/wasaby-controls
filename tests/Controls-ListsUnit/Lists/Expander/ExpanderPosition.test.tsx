/**
 * @jest-environment jsdom
 */
import { unmountComponentAtNode } from 'react-dom';
import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { WasabyEvents } from 'UICore/Events';

import { renderTree } from 'Controls-ListsUnit/Lists/TreeHelpers';
import { getCell } from 'Controls-ListsUnit/Lists/GridHelpers';

import ExpanderPosition from 'Controls-demo/treeGridReact/Expander/ExpanderPosition';

describe('Controls-ListsUnit/Lists/Expander/ExpanderPosition', () => {
    let container = null;

    beforeEach(() => {
        container = document.createElement('div');
        WasabyEvents.initInstance(container);
        document.body.appendChild(container);
    });

    afterEach(() => {
        unmountComponentAtNode(container);
        WasabyEvents.destroyInstance(container);
        container.remove();
        container = null;
    });

    it('expander is displayed left if expanderPosition=default', () => {
        const { getFirstItem } = renderTree(
            <ExpanderPosition expanderPosition={'default'} />,
            { container }
        );
        const firstColumn = getCell(getFirstItem(), 0);
        expect(firstColumn).toMatchSnapshot();
    });

    it('expander is displayed right if expanderPosition=right', () => {
        const { getFirstItem } = renderTree(
            <ExpanderPosition expanderPosition={'right'} />,
            { container }
        );
        const firstColumn = getCell(getFirstItem(), 0);
        expect(firstColumn).toMatchSnapshot();
    });

    it('should rerender if change expanderPosition', async () => {
        const { getFirstItem } = renderTree(
            <ExpanderPosition expanderPosition={'default'} />,
            { container }
        );

        const buttonWithRightValue = screen.queryByText('right');
        await userEvent.click(buttonWithRightValue);

        await waitFor(() => {
            const firstColumn = getCell(getFirstItem(), 0);
            // Экспандер должен быть справа
            expect(firstColumn).toMatchSnapshot();
        });
    });
});
