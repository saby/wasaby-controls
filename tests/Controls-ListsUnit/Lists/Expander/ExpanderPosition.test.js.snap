// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls-ListsUnit/Lists/Expander/ExpanderPosition expander is displayed left if expanderPosition=default 1`] = `
<div
  class="controls-GridReact__cell tw-flex tw-relative js-controls-Grid__row-cell controls-GridReact__minHeight-data controls-fontsize-2xl tw-cursor-pointer controls-padding_top-grid_s controls-padding_bottom-grid_s controls-padding_left-m controls-padding_right-grid_m controls-hover-background-list_default tw-items-baseline controls-GridReact__cell-baseline tw-break-words"
>
  <div
    class="tw-flex tw-items-baseline "
  >
    <div
      class="js-controls-Tree__row-expander controls-TreeGrid__row-expander controls-cursor_pointer js-controls-ListView__notEditable controls-TreeGrid__row_default-expander_size_default tw-inline-flex tw-justify-center tw-items-center tw-flex-shrink-0 controls-TreeGrid__row-expander_node_default_position_default controls-TreeGrid__row-expander_node_default controls-TreeGrid__row-expander_node_iconSize_default controls-TreeGrid__row-expander_node_iconStyle_default controls-TreeGrid__row-expander_collapsed controls-TreeGrid__row-expander_node_default_collapsed"
      data-qa="row-expander"
      tabindex="-1"
    />
  </div>
  Apple
</div>
`;

exports[`Controls-ListsUnit/Lists/Expander/ExpanderPosition expander is displayed right if expanderPosition=right 1`] = `
<div
  class="controls-GridReact__cell tw-flex tw-relative js-controls-Grid__row-cell controls-GridReact__minHeight-data controls-fontsize-2xl tw-cursor-pointer controls-padding_top-grid_s controls-padding_bottom-grid_s controls-padding_left-m controls-padding_right-grid_m controls-hover-background-list_default tw-items-baseline controls-GridReact__cell-baseline tw-break-words"
>
  <div
    class="tw-flex tw-items-baseline "
  />
  Apple
  <div
    class="js-controls-Tree__row-expander controls-TreeGrid__row-expander controls-cursor_pointer js-controls-ListView__notEditable tw-inline-block controls-padding_left-xs controls-TreeGrid__row-expander_node_default_position_right controls-TreeGrid__row-expander_node_default controls-TreeGrid__row-expander_node_iconSize_default controls-TreeGrid__row-expander_node_iconStyle_default controls-TreeGrid__row-expander_collapsed controls-TreeGrid__row-expander_node_default_collapsed"
    data-qa="row-expander"
    tabindex="-1"
  />
</div>
`;

exports[`Controls-ListsUnit/Lists/Expander/ExpanderPosition should rerender if change expanderPosition 1`] = `
<div
  class="controls-GridReact__cell tw-flex tw-relative js-controls-Grid__row-cell controls-GridReact__minHeight-data controls-fontsize-2xl tw-cursor-pointer controls-padding_top-grid_s controls-padding_bottom-grid_s controls-padding_left-m controls-padding_right-grid_m controls-hover-background-list_default tw-items-baseline controls-GridReact__cell-baseline tw-break-words"
>
  <div
    class="tw-flex tw-items-baseline "
  />
  Apple
  <div
    class="js-controls-Tree__row-expander controls-TreeGrid__row-expander controls-cursor_pointer js-controls-ListView__notEditable tw-inline-block controls-padding_left-xs controls-TreeGrid__row-expander_node_default_position_right controls-TreeGrid__row-expander_node_default controls-TreeGrid__row-expander_node_iconSize_default controls-TreeGrid__row-expander_node_iconStyle_default controls-TreeGrid__row-expander_collapsed controls-TreeGrid__row-expander_node_default_collapsed"
    data-qa="row-expander"
    tabindex="-1"
  />
</div>
`;
